File: notebookspython/21-queue_with_stacks.py
"""
实现说明：
通过两个栈 (in_stack, out_stack) 模拟队列的先进先出 (FIFO) 特性。
核心逻辑是“延迟搬运”：仅当 out_stack 为空时，才将 in_stack 的数据
倒序压入 out_stack，从而实现 O(1) 的均摊时间复杂度。
"""
class MyQueue:
    def __init__(self):
        # 使用 Python 的 list 作为栈
        self.in_stack = []
        self.out_stack = []

    def push(self, x: int) -> None:
        self.in_stack.append(x)

    def pop(self) -> int:
        self.peek()  # 确保 out_stack 有数据
        return self.out_stack.pop()

    def peek(self) -> int:
        # 只有当 out_stack 为空时，才将 in_stack 的数据倒过来
        if not self.out_stack:
            while self.in_stack:
                self.out_stack.append(self.in_stack.pop())
        return self.out_stack[-1]

    def empty(self) -> bool:
        return not self.in_stack and not self.out_stack

File: notebooksjava/_21QueueWithStacks.java
import java.util.ArrayDeque;
import java.util.Deque;

public class _21QueueWithStacks {

    static class MyQueue {
        private final Deque<Integer> inStack = new ArrayDeque<>();
        private final Deque<Integer> outStack = new ArrayDeque<>();

        public void push(int x) {
            inStack.push(x);
        }

        public int pop() {
            peek();
            return outStack.pop();
        }

        public int peek() {
            if (outStack.isEmpty()) {
                while (!inStack.isEmpty()) {
                    outStack.push(inStack.pop());
                }
            }
            return outStack.peek();
        }

        public boolean empty() {
            return inStack.isEmpty() && outStack.isEmpty();
        }
    }

    public static void main(String[] args) {
        MyQueue queue = new MyQueue();
        queue.push(1);
        queue.push(2);

        if (queue.peek() != 1) {
            throw new AssertionError("Peek failed");
        }
        if (queue.pop() != 1) {
            throw new AssertionError("Pop failed");
        }
        if (queue.pop() != 2) {
            throw new AssertionError("Second pop failed");
        }
        if (!queue.empty()) {
            throw new AssertionError("Queue should be empty");
        }

        System.out.println("All test cases passed ✅");
    }
}
